src/internal/consistenthashing/etcd.go

```diff
@@ -0,0 +1,370 @@
+			}); err != nil {
+			ring.Logger.WithFields(ring.memberLogFields()).Errorf("failed to renew etcd lease: %v", err)
+
+			nodeReady <- errors.EnsureStack(err)
+		}
+	}()
+
+	return <-nodeReady
+}
+
+// watch watches etcd keys with prefix Ring.Prefix to determine if a node has been added or removed.
+// When a node is added, the ring calls rebalance to release any locks that no longer associate to its node. Rebalance
+// on delete happens by default since each Member attempts to get all locks. When a Member is deleted,
+// the pending calls to Lock by each Member will go through on the nodes that associates to given lock.
+func (ring *Ring) watch(col collection.EtcdCollection) {
+	if err := col.ReadOnly(ring.Ctx).WatchF(func(event *watch.Event) error {
+		nodeKey := string(event.Key)
+		switch event.Type {
+		case watch.EventDelete:
+			ring.removeMemberById(ring.etcdToMember(nodeKey))
+
+		case watch.EventPut:
+			ring.insertMemberById(ring.etcdToMember(nodeKey))
+
+			// Need to release locks that no longer associate to the ring's node.
+			if err := ring.rebalance(); err != nil {
+				ring.Logger.WithField("ring", ring.Prefix).Errorf("failed rebalancing: %v", err)
+				return err
+			}
+		}
+
+		return nil
+	}); err != nil {
+		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+			ring.Logger.WithFields(ring.memberLogFields()).Infof("shutting down ring")
+			return
+		}
+
+		ring.Logger.WithFields(ring.memberLogFields()).Errorf("failed watch: %v", err)
+	}
+}
+
+func (ring *Ring) rebalance() error {
+	var errs []error
+	ring.node.Locks.Range(func(k, v any) bool {
+		key := k.(string)
+		if ring.Get(key).Id == ring.node.Id {
+			return true // return true acts as continue in a sync.Map.Range()
+		}
+
+		lockInfo := v.(lockInfo)
+		if err := lockInfo.Lock.Unlock(lockInfo.Ctx); err != nil {
+			errs = append(errs, errors.EnsureStack(err))
+		}
+
+		ring.node.Locks.Delete(key)
+
+		ring.Logger.WithFields(ring.lockLogFields(key)).Info("rebalanced lock")
+
+		return true
+	})
+
+	if errs != nil {
+		return errors.New(fmt.Sprintf("errors unlocking locks while rebalancing: %v", errs))
+	}
+
+	return nil
+}
+
+func defaultRing(ctx context.Context, prefix string, client *etcd.Client, logger *logrus.Logger) *Ring {
+	ctx, cancel := context.WithCancel(ctx)
+
+	ring := &Ring{
+		Client: client,
+		Prefix: prefix,
+		Logger: logger,
+
+		Ctx:    ctx,
+		Cancel: cancel,
+
+		hashFn: crc32.ChecksumIEEE,
+	}
+
+	return ring
+}
+
+// insertMemberById is a convenience wrapper for insertMember that constructs the Member on behalf of the caller.
+func (ring *Ring) insertMemberById(id string) {
+	member := Member{
+		Id:   id,
+		Hash: ring.hashFn([]byte(id)),
+	}
+
+	ring.insertMember(member)
+}
+
+// insertMember updates Ring.members, and sorts Ring.members by the Member hash.
+// If the Member already exists, insertMember is a no-op. The member can be mocked for testing purposes.
+func (ring *Ring) insertMember(member Member) {
```

Comment:
I think we need to lock modification to the members list. A concurrent lock could be problematic.